lib/common/bitstream.h

```diff
@@ -332,7 +332,16 @@ MEM_STATIC FORCE_INLINE_ATTR size_t BIT_getMiddleBits(size_t bitContainer, U32 c
     U32 const regMask = sizeof(bitContainer)*8 - 1;
     /* if start > regMask, bitstream is corrupted, and result is undefined */
     assert(nbBits < BIT_MASK_SIZE);
+    /* x86 transform & ((1 << nbBits) - 1) to bzhi instruction, it is better
+     * than accessing memory. When bmi2 instruction is not present, we consider
+     * such cpus old (pre-Haswell, 2013) and their performance is not of that
+     * importance.
+     */
+#if defined(__x86_64__) || defined(_M_X86)
+    return (bitContainer >> (start & regMask)) & ((((U64)1) << nbBits) - 1);
+#else
     return (bitContainer >> (start & regMask)) & BIT_mask[nbBits];
+#endif
```

Comment:
Doesn't this make testing across archs and evaluating correctness harder? Wouldn't it be better to define a macro that's equivalent to what `BIT_mask[nBits]` gives you, at least?

Reply:
Given zstd has CIs for many architectures, I don't think this can be considered as a blocker.

Macro that computes BIT_mask[nBits] is worse for ARM because of memory cache accesses, accessing BIT_mask is fast because it fits cache very well, computing `(1 << x) - 1` requires 2-3 extra instructions. BMI2 x86 uses bzhi instruction which is clearly a better decision. If your suggestion is to move BIT_mask to a macro for x86, I can do that, not sure if that will make code readable

I understand the frustration behind such decisions. However, as zstd is a performance critical piece of software, such decisions already occur here and there

Reply:
> If your suggestion is to move BIT_mask to a macro for x86, I can do that, not sure if that will make code more readable

Yes, that is my suggestion. If you define it close to the look up table, it helps in understanding what's going on and will make it simpler to be used elsewhere, if someone wants to.